File: C#/Engine.Core/Maths/Set.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Engine.Core.Maths
{
    public class Set<T>
    {
        protected List<Interval> intervals;

        /// <summary>
        /// only used if for example it intersects with the natural numbers
        /// </summary>
        protected List<T> containingValues = null;

        public Set(params T[] values)
        {
            if (values != null)
            {
                containingValues = new List<T>(values);
            }
            if (typeof(T) == typeof(float)) intervals = new List<Interval>();
        }

        /// <summary>
        /// Calculating the cardinality of the interval
        /// </summary>
        /// <returns>Returns the cardinality. If the value is infinty it return -1 instead.</returns>
        public virtual int Cardinality()
        {
            if (containingValues == null) return -1;
            else
            {
                return containingValues.Count;
            }
        }

        public virtual bool Contains(T value)
        {
            return containingValues?.Contains(value) ?? false;
        }

        public void AddInterval(Interval interval)
        {
            if (typeof(T) == typeof(float))
                intervals.Add(interval);
            else Debug.Fail("This should be a float-set to add an interval");
        }
    }
}

File: Engine/src/Core/Maths/Set.cpp
#include <Set.h>

namespace Engine::Core::Maths
{
    template <typename T>
    Set<T>::Set(float *ptr = nullptr)
    {
        containingValues = NULL;
    }

    template <typename T>
    Set<T>::Set(std::initializer_list<T> values)
    {
        if (!values.empty())
        {
            containingValues.assign(values);
        }
        if constexpr (std::is_same<T, float>::value)
        {
            intervals = std::vector<Interval>();
        }
    }

    template <typename T>
    int Set<T>::Cardinality() const
    {
        if (containingValues.empty())
        {
            return -1;
        }
        return containingValues.size();
    }

    template <typename T>
    bool Set<T>::Contains(const T &value) const
    {
        return !containingValues.empty() && std::find(containingValues.begin(), containingValues.end(), value) != containingValues.end();
    }

    template <typename T>
    void Set<T>::AddInterval(const Interval &interval)
    {
        if constexpr (std::is_same<T, float>::value)
        {
            intervals.push_back(interval);
        }
        else
        {
            std::cerr << "This should be a float-set to add an interval" << std::endl;
        }
    }
}